### PR TITLE
Addition of contributing guide

### PR DESCRIPTION
**Pull request by bijington**
## Description

Adds a contributing guidelines document largely inspired by [https://github.com/xamarin/XamarinCommunityToolkit/blob/main/CONTRIBUTING.md](https://github.com/xamarin/XamarinCommunityToolkit/blob/main/CONTRIBUTING.md)

### CODE REVIEW

File: CONTRIBUTING.md
Comment (bijington) [inline CONTRIBUTING.md:144]: This is from XCT but it appears we have a inconsistent approach already in MCT:

* `_` prefix -> https://github.com/CommunityToolkit/Maui/blob/00f8f8de9e81423a43800404c1d6ba932f8a7690/src/CommunityToolkit.Maui/Alerts/Snackbar/SnackBar.shared.cs#L11
* `camelCaseFieldName` -> https://github.com/CommunityToolkit/Maui/blob/00f8f8de9e81423a43800404c1d6ba932f8a7690/src/CommunityToolkit.Maui/Behaviors/BaseBehavior.shared.cs#L21

File: CONTRIBUTING.md
Comment (bijington) [inline CONTRIBUTING.md:1]: TODO: add text and flowchart diagram indicating where changes should be located based on #214

**Comment by bijington** [comment]
Also investigate `.editorconfig` changes to make it easier to follow with the guidelines

**Comment by brminnick** [comment]
+1 to `.editorconfig` rules that'll automatically enforce these 💯 

Can you also add not to use the [Null-Forgiving Operator, `!`](https://docs.microsoft.com/dotnet/csharp/language-reference/operators/null-forgiving).

**Comment by VladislavAntonyuk** [comment]
> +1 to `.editorconfig` rules that'll automatically enforce these 💯
> 
> Can you also add not to use the [Null-Forgiving Operator, `!`](https://docs.microsoft.com/dotnet/csharp/language-reference/operators/null-forgiving).

I have one. We can use it as start point:
[editorconfig.zip](https://github.com/CommunityToolkit/Maui/files/7707474/editorconfig.zip)

File: CONTRIBUTING.md
Comment (VladislavAntonyuk) [reply CONTRIBUTING.md:144]: can we fix it in https://github.com/CommunityToolkit/Maui/pull/214 or it will be a new PR with editorConfig rules?

**Comment by bijington** [comment]
> Can you also add not to use the [Null-Forgiving Operator, `!`](https://docs.microsoft.com/dotnet/csharp/language-reference/operators/null-forgiving). 

There is a bullet point in the guidelines stating we should avoid using it. I will see if we can prevent its use entirely with .editorconfig

**Comment by bijington** [comment]
> > +1 to `.editorconfig` rules that'll automatically enforce these 💯
> 
> > 
> 
> > Can you also add not to use the [Null-Forgiving Operator, `!`](https://docs.microsoft.com/dotnet/csharp/language-reference/operators/null-forgiving).
> 
> 
> 
> I have one. We can use it as start point:
> 
> [editorconfig.zip](https://github.com/CommunityToolkit/Maui/files/7707474/editorconfig.zip)
> 
> 
> 
> 

That's really helpful thank you!

File: CONTRIBUTING.md
Comment (bijington) [reply CONTRIBUTING.md:144]: I don't mind. If #214 is near completion I'm happy to include in this or whichever PR includes the .editorconfig addition.

I guess we should decide which approach we prefer first 😃

Comment (brminnick) [reply CONTRIBUTING.md:144]: Let's make those changes here in this PR.

#214 is a massive PR and just got 2 Approvals. I'm exhausted just thinking about reviewing it again 😅

Comment (brminnick) [reply CONTRIBUTING.md:144]: For what it's worth, I prefer prefixing fields with `_`.

That's just my personal preference (makes it easy to determine at a glance whether a variable's scope is local or global). It's not a hill I'm willing to die on, though, if everyone else prefers to nix the `_` prefix.

Comment (bijington) [reply CONTRIBUTING.md:144]: I lean the other way but again I'm not going to fight to the death on it. If it means we have fewer differences to the linked standard from dotnet foundation then that's probably a better approach. To clarify this would be in favour of the `_` prefix

File: src/CommunityToolkit.Maui/Alerts/Snackbar/Snackbar.android.cs
Comment (VladislavAntonyuk) [inline src/CommunityToolkit.Maui/Alerts/Snackbar/Snackbar.android.cs:70]: it would be nice if we define the style for `if` blocks. somewhere we use braces (line 70), somewhere we skip them (line 64).

**Comment by VladislavAntonyuk** [comment]
what do you think about `dotnet format` in git pre-commit hooks? it will automatically validate and fix rules

**Comment by brminnick** [comment]
> what do you think about dotnet format in git pre-commit hooks? it will automatically validate and fix rules

Love it! How do we do that?

**Comment by VladislavAntonyuk** [comment]
> > what do you think about dotnet format in git pre-commit hooks? it will automatically validate and fix rules
> 
> Love it! How do we do that?

https://vladislavantonyuk.azurewebsites.net/articles/How-to-automatically-keep-your-code-clean

**Comment by bijington** [comment]
> > > what do you think about dotnet format in git pre-commit hooks? it will automatically validate and fix rules
> > 
> > Love it! How do we do that?
> 
> https://vladislavantonyuk.azurewebsites.net/articles/How-to-automatically-keep-your-code-clean

I like the idea as it means we shouldn't need the bot that does the same thing and it should result in fewer PRs to review.

Is there any risk in it changing things so much from what the author of the commit originally wrote that it may confuse things?

File: src/CommunityToolkit.Maui/Alerts/Snackbar/Snackbar.android.cs
Comment (bijington) [reply src/CommunityToolkit.Maui/Alerts/Snackbar/Snackbar.android.cs:70]: Definitely! Great spot! Time for another vote I think

**Comment by VladislavAntonyuk** [comment]
> > > > what do you think about dotnet format in git pre-commit hooks? it will automatically validate and fix rules
> > > 
> > > 
> > > Love it! How do we do that?
> > 
> > 
> > https://vladislavantonyuk.azurewebsites.net/articles/How-to-automatically-keep-your-code-clean
> 
> I like the idea as it means we shouldn't need the bot that does the same thing and it should result in fewer PRs to review.
> 
> Is there any risk in it changing things so much from what the author of the commit originally wrote that it may confuse things?

We can describe it in the guide (with the link to the issue/discussion) but implement it a bit later.

**Comment by bijington** [comment]
I believe this is ready for review. I have currently changed to encourage/force:

- camelCaseNaming
- `is null` over `== null`
- `{ }` after `if`, `for`, etc.

I have not yet managed to get the `!` null forgiving operator banned. This package https://github.com/tom-englert/Nullable.Extended allows us to but the correcting behaviour is to provide a commit to omit it rather than remove it which feels wrong for this use. Thoughts?

**Review by VladislavAntonyuk — APPROVED** [review]

File: .editorconfig
Comment (brminnick) [inline .editorconfig:95]: Are we missing something after the equals sign? Or does leaving it blank ensure that we don't use `_` for fields?

Comment (bijington) [reply .editorconfig:95]: This is left blank to prevent the _ prefix as you suggest

**Pull request merged**
